onnxruntime/contrib_ops/cuda/transformers/beam_search.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cuda_execution_provider.h"
+#include "contrib_ops/cuda/transformers/beam_search.h"
+#include "beam_search_device_helper.h"
+#include "dump_cuda_tensor.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    BeamSearch,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    (*KernelDefBuilder::Create())
+        .InputMemoryType(OrtMemTypeCPUInput, 0)    // 'input_ids' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 1)    // 'max_length' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 2)    // 'min_length' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 3)    // 'num_beams' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 4)    // 'num_return_sequences' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 5)    // 'temperature' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 6)    // 'length_penalty' needs to be on CPU
+        .InputMemoryType(OrtMemTypeCPUInput, 7)    // 'repetition_penalty' needs to be on CPU
+        .OutputMemoryType(OrtMemTypeCPUOutput, 0)  // 'sequences' output on CPU
+        .OutputMemoryType(OrtMemTypeCPUOutput, 1)  // 'sequences_scores' output on CPU
+        .TypeConstraint("T", {DataTypeImpl::GetTensorType<float>(),
+                              DataTypeImpl::GetTensorType<MLFloat16>()}),
+    BeamSearch);
+
+transformers::CudaTensorConsoleDumper g_cuda_dumper;
+
+BeamSearch::BeamSearch(const OpKernelInfo& info)
+    : onnxruntime::contrib::transformers::BeamSearch(info) {
+  SetComputeStream(static_cast<void*>(info.GetExecutionProvider()->GetComputeStream()));
+
+  SetDeviceHelpers(BeamSearchCudaDeviceHelper::AddToFeeds,
+                   BeamSearchCudaDeviceHelper::TopK);
+
+  SetDeviceHelpers(BeamSearchCudaDeviceHelper::ProcessLogits<float>,
+                   BeamSearchCudaDeviceHelper::InitBeamState<float>,
+                   BeamSearchCudaDeviceHelper::DeviceCopy<float>,
+                   BeamSearchCudaDeviceHelper::UpdateFeeds<float>);
+
+  SetDeviceHelpers(BeamSearchCudaDeviceHelper::ProcessLogits<MLFloat16>,
+                   BeamSearchCudaDeviceHelper::InitBeamState<MLFloat16>,
+                   BeamSearchCudaDeviceHelper::UpdateFeeds<MLFloat16>);
+
+  SetConsoleDumper(&g_cuda_dumper);
+}
+
+Status BeamSearch::ComputeInternal(OpKernelContext* context) const {
+  return onnxruntime::contrib::transformers::BeamSearch::Compute(context);
```

Comment:
So it calls cpu kernel and part of the impls are on cuda controlled by devicehelper. May I know the benefit of design this way?

Reply:
For beam search, some code (like subgraph execution) can be shared between CPU and CUDA. Device helpers are device specified implementation. It is quite similar to virtual function but better in isolating CPU and CUDA implementations.